projects/ngneat/error-tailor/src/lib/control-error.directive.ts

```diff
@@ -144,12 +154,30 @@ export class ControlErrorsDirective implements OnInit, OnDestroy {
       .pipe(takeUntil(this.destroy))
       .subscribe(() => {
         const hasErrors = !!this.control.errors;
+
         if (hasErrors) {
           this.showError();
         } else {
           this.hideError();
         }
       });
+
+    changesOnTouched$.pipe(takeUntil(this.destroy)).subscribe(() => {
+      const hasErrors = !!this.control.errors;
+      const touched = this.mergedConfig.controlErrorsOn.touched ? this.control.touched : true;
+
+      if (hasErrors && touched) {
+        this.showError();
+      } else {
+        this.hideError();
+      }
+    });
+  }
+
+  ngDoCheck(): void {
```

Comment:
Let's use afterRender hook and only invoke it when this.mergedConfig.controlErrorsOn.touched

Reply:
What about monkey patch ngControl methods instead?

Reply:
I considered monkey patching since it would be the most performant solution (called the least amount of times), but I think it's not the safest solution.

First of all we'd need to monkey patch all methods that might change the `touched` property, so `control.markAsTouched()`, `control.markAsUntouched()`, `formGroup.markAllAsTouched()`. We also could miss methods introduced in new versions of Angular that change that property, for example `formGroup.markAllAsTouched()` was introduced in v8.

Reply:
Yes, because it'll only do it if this.mergedConfig.controlErrorsOn.touched is true

Reply:
So, I tried the `afterRender` implementation as you suggested by adding this to the `ngOnInit()`:

```ts
if (this.mergedConfig.controlErrorsOn.touched) {
  afterRender(
    () => {
      this.touchedChanges$.next(this.control.touched);
    },
    { inject: this.injector }
  );
}
```

Is this what you meant? If so, the problem with this solution is that the `afterRender` callback runs after change detection has already run, so the template will be out of sync until change detection runs again, as shown here:

![CleanShot 2024-05-29 at 12 18 00](https://github.com/ngneat/error-tailor/assets/677873/f43200c9-5142-4737-bfc0-f7683c8ac60a)

To fix this I would need to call `changeDetectorRef.detectChanges()` which will then run change detection again and update the template:

```ts
if (this.mergedConfig.controlErrorsOn.touched) {
  afterRender(
    () => {
      this.touchedChanges$.next(this.control.touched);
      this.changeDetectorRef.detectChanges();
    },
    { inject: this.injector }
  );
}
```

I wonder if this is any better than having the `ngDoCheck` that short circuits when `this.mergedConfig.controlErrorsOn.touched` is false, in terms of performance.

What do you think?

Reply:
No, it's not. My issue with doCheck is that most of the users including me don't need this functionality and don't need this extra redundant call run

Reply:
What about observing the ng-touched class?

Reply:
That's another possible solution! Before implementing this I did quickly test an implementation using `MutationObserver` and it should be possible. 

It feels a bit less "the Angular way" and relies on a class being present in the DOM, instead of the value of a property on an `AbstractControl` instance. Which might not be a problem per se, since I think we can trust that the `ng-touched` class is not going to be changed/removed anytime soon.

But if you're after a performance improvement I wonder how much would we gain. And hopefully with this solution we don't need to call `changeDetectorRef.detectChanges()` as well.

Reply:
Still I can give it a try if you want!

Reply:
We'll gain that people that doesn't need it will not run a redundant call

Reply:
Ok, let me give it a try and I will let you know the results 👍

Reply:
To align on the purpose of this, can you confirm that the final goal should be to cause the least amount of rerenders of the component?

Reply:
Yes, and to only run code when someone needs it

Reply:
Back with some results. I compared these three implementations:

- `ngDoCheck` (as in the current PR):
  ```ts
  ngDoCheck(): void {
    if (this.mergedConfig.controlErrorsOn.touched) {
      this.touchedChanges$.next(this.control.touched);
    }
  }
  ```
- `afterRender`:
  ```ts
  
  if (this.mergedConfig.controlErrorsOn.touched) {
    afterRender(
      () => {
        this.touchedChanges$.next(this.control.touched);
      },
      { injector: this.injector },
    );
  }
  ```
  and then call `this.changeDetectorRef.detectChanges()`
- `MutationObserver`:
  ```ts
  if (this.mergedConfig.controlErrorsOn.touched) {
    const observer = new MutationObserver((mutations: MutationRecord[]) => {
      mutations.forEach((mutation) => {
        const isTouched = (mutation.target as HTMLElement).classList.contains('ng-touched');
        this.touchedChanges$.next(isTouched);
      });
    });
    observer.observe(this.host, {
      attributeFilter: ['class'],
    });
  }
  ```

For each implementation I placed a `console.count('rerender')` in `ngDoCheck()` to count how many rerenders were performed (only for the name form control). I tested by loading the page, then after one second I called `markAsTouched()` and then after another second I called `markAsUntouched()`. If I didn't do any mistake, these are the results:

- `ngDoCheck`: 5 rerenders
  <details>
    <summary>GIF</summary>
   
    ![ngdocheck](https://github.com/ngneat/error-tailor/assets/677873/6f97c0b5-2b8f-491e-8657-f442a4a1fd4e)

  </details>
- `afterRender`: 7 rerenders
  <details>
    <summary>GIF</summary>
    
    ![afterrender](https://github.com/ngneat/error-tailor/assets/677873/0bec72c2-9f2a-4f7c-858e-c8e6efc5eebd)

  </details>
- `MutationObserver`: 8 rerenders
  <details>
    <summary>GIF</summary>
    
    ![mutationobserver](https://github.com/ngneat/error-tailor/assets/677873/900cab8c-f991-4133-bbd6-adc54cd06018)

  </details>

Honestly I don't understand why the `MutationObserver` implementation would cause 3 more rerenders. If you want I can push the code to a separate branch on my fork if you want to take a look, let me know. 

If these results are correct then maybe it's better to go with the `ngDoCheck` implementation? As I see it right now the comparison is between:
- `ngDoCheck` solution: accessing the `this.mergedConfig.controlErrorsOn.touched` on each change detection (doesn't do anything else if false)
- other solutions: causing more rerenders, which will definitely run more code since the whole Angular's change detection logic will be set into motion.

What do you think? Had some fun digging into this matter 😄 but if you feel strongly that we should go for the other solutions I can implement the one you prefer.

Reply:
Let's go with afterRender but only call detect changes if the prev !== current

Reply:
Sure, going to call `detectChanges` in the `changesOnTouched$` which has `distinctUntilChanged`, so will only run when it actually changes.

Reply:
I'm very busy at work currently. Try to comment some code and see where is the issue (You can maybe begin with setError)

Reply:
@NetanelBasal Does it even make sense to use `afterRender`? Something that will need for sure a render to be shown, probably doesn't make sense to be in `afterRender`.

Furthermore, the documentation says that
> The execution of render callbacks are not tied to any specific component instance, but instead an application-wide hook.
so the callback would be run after every render application-wide.

Reply:
> so the callback would be run after every render application-wide.

Correct. That's one of the main issues with this approach. I think we can go with MutationObserver or monkey patching